File: golang/linkedlist/reverse.go
// reverse a linked list

package main

import "fmt"

type Node struct {
	data int
	next *Node
}

func reverse(head *Node) {
	var prev *Node
	current := head
	for current != nil {
		next := current.next
		current.next = prev
		prev = current
		current = next
	}
	head = prev
}

func reverseRecursion(head *Node) *Node {
	if head == nil || head.next == nil {
		return head
	}
	rest := reverseRecursion(head.next)
	head.next.next = head
	head.next = nil
	return rest
}

func printList(head *Node) {
	for head != nil {
		fmt.Print(head.data, " ")
		head = head.next
	}
	fmt.Println()
}

func main() {
	head := &Node{data: 1}
	head.next = &Node{data: 2}
	head.next.next = &Node{data: 3}
	head.next.next.next = &Node{data: 4}
	head.next.next.next.next = &Node{data: 5}
	head.next.next.next.next.next = &Node{data: 6}

	fmt.Println("Original linked list")
	printList(head)

	reverse(head)
	fmt.Println("Reversed linked list")
	printList(head)

}

File: Questions/questions/linkedList/Reverse.java

public class Reverse {

    private Node head;

    static class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            next = null;

        }
    }

    Node reverse(Node node) {

        Node prev = null;
        Node current = node;
        Node next = null;

        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        node = prev;
        return node;
    }

    Node reverseRecursion(Node node) {
        if (node == null || node.next == null) {
            return node;
        }
        Node rest = reverseRecursion(node.next);
        node.next.next = node;
        node.next = null;
        return rest;
    }

    void printList(Node node) {
        while (node != null) {
            System.out.print(node.data + " ");
            node = node.next;
        }
    }

    public static void main(String[] args) {

        Reverse list = new Reverse();
        list.head = new Node(1);
        list.head.next = new Node(2);
        list.head.next.next = new Node(3);
        list.head.next.next.next = new Node(4);
        list.head.next.next.next.next = new Node(5);

        System.out.println("Given Linked list");
        list.printList(list.head);
        // list.head = list.reverse(list.head);
        list.head = list.reverseRecursion(list.head);
        System.out.println("");
        System.out.println("Reversed linked list ");
        list.printList(list.head);
    }

}
